src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/tools/monitoring/outputMonitor.ts

```diff
@@ -0,0 +1,254 @@
+ *--------------------------------------------------------------------------------------------*/
+
+import { timeout } from '../../../../../../../base/common/async.js';
+import { CancellationToken } from '../../../../../../../base/common/cancellation.js';
+import { Emitter, Event } from '../../../../../../../base/common/event.js';
+import { MarkdownString } from '../../../../../../../base/common/htmlContent.js';
+import { Disposable } from '../../../../../../../base/common/lifecycle.js';
+import { localize } from '../../../../../../../nls.js';
+import { ExtensionIdentifier } from '../../../../../../../platform/extensions/common/extensions.js';
+import { IChatWidgetService } from '../../../../../chat/browser/chat.js';
+import { ChatElicitationRequestPart } from '../../../../../chat/browser/chatElicitationRequestPart.js';
+import { ChatModel } from '../../../../../chat/common/chatModel.js';
+import { IChatService } from '../../../../../chat/common/chatService.js';
+import { ILanguageModelsService, ChatMessageRole } from '../../../../../chat/common/languageModels.js';
+import { IToolInvocationContext } from '../../../../../chat/common/languageModelToolsService.js';
+import { ITaskService } from '../../../../../tasks/common/taskService.js';
+import { PollingConsts } from '../../bufferOutputPolling.js';
+import { IPollingResult, OutputMonitorState, IExecution, IRacePollingOrPromptResult } from './types.js';
+import { getResponseFromStream } from './utils.js';
+
+export interface IOutputMonitor extends Disposable {
+	readonly isIdle: boolean;
+
+	readonly onDidFinishCommand: Event<void>;
+	readonly onDidIdle: Event<void>;
+	readonly onDidTimeout: Event<void>;
+
+	startMonitoring(
+		chatService: IChatService,
+		command: string,
+		invocationContext: any,
+		token: CancellationToken
+	): Promise<IPollingResult>;
+}
+
+export class OutputMonitor extends Disposable implements IOutputMonitor {
+	private _isIdle = false;
+	private _state: OutputMonitorState = OutputMonitorState.Initial;
+	get state(): OutputMonitorState {
+		return this._state;
+	}
+
+	private readonly _onDidFinishCommand = this._register(new Emitter<void>());
+	readonly onDidFinishCommand = this._onDidFinishCommand.event;
+	private readonly _onDidIdle = this._register(new Emitter<void>());
+	readonly onDidIdle = this._onDidIdle.event;
+	private readonly _onDidTimeout = this._register(new Emitter<void>());
+	readonly onDidTimeout = this._onDidTimeout.event;
+
+	get isIdle(): boolean {
+		return this._isIdle;
+	}
+
+	constructor(
+		private readonly _execution: IExecution,
+		@ILanguageModelsService private readonly _languageModelsService: ILanguageModelsService,
+		@ITaskService private readonly _taskService: ITaskService,
+		@IChatWidgetService private readonly _chatWidgetService: IChatWidgetService,
+		private readonly _pollFn?: (execution: IExecution, token: CancellationToken, taskService: ITaskService) => Promise<IPollingResult | undefined> | undefined,
```

Comment:
Having this as last arg breaks dependency injection, removing the `?` and move it below `_execution` to fix